packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts

```diff
@@ -0,0 +1,275 @@
+  [AST_NODE_TYPES.TSParenthesizedType]: ts.SyntaxKind.ParenthesizedType;
+  [AST_NODE_TYPES.TSPropertySignature]: ts.SyntaxKind.PropertySignature;
+  [AST_NODE_TYPES.TSPublicKeyword]: ts.SyntaxKind.PublicKeyword;
+  [AST_NODE_TYPES.TSPrivateKeyword]: ts.SyntaxKind.PrivateKeyword;
+  [AST_NODE_TYPES.TSProtectedKeyword]: ts.SyntaxKind.ProtectedKeyword;
+  [AST_NODE_TYPES.TSQualifiedName]: ts.SyntaxKind.QualifiedName;
+  [AST_NODE_TYPES.TSReadonlyKeyword]: ts.SyntaxKind.ReadonlyKeyword;
+  [AST_NODE_TYPES.TSRestType]: ts.SyntaxKind.RestType;
+  [AST_NODE_TYPES.TSStaticKeyword]: ts.SyntaxKind.StaticKeyword;
+  [AST_NODE_TYPES.TSStringKeyword]: ts.SyntaxKind.StringKeyword;
+  [AST_NODE_TYPES.TSSymbolKeyword]: ts.SyntaxKind.SymbolKeyword;
+  [AST_NODE_TYPES.TSThisType]: ts.SyntaxKind.ThisType;
+  [AST_NODE_TYPES.TSTupleType]: ts.SyntaxKind.TupleType;
+  [AST_NODE_TYPES.TSTypeAliasDeclaration]: ts.SyntaxKind.TypeAliasDeclaration;
+  [AST_NODE_TYPES.TSTypeAnnotation]: undefined;
+  [AST_NODE_TYPES.TSTypeAssertion]: ts.SyntaxKind.TypeAssertionExpression;
+  [AST_NODE_TYPES.TSTypeLiteral]: ts.SyntaxKind.TypeLiteral;
+  [AST_NODE_TYPES.TSTypeOperator]: ts.SyntaxKind.TypeOperator;
+  [AST_NODE_TYPES.TSTypeParameter]: ts.SyntaxKind.TypeParameter;
+  [AST_NODE_TYPES.TSTypeParameterDeclaration]: undefined;
+  [AST_NODE_TYPES.TSTypeParameterInstantiation]:
+    | ts.SyntaxKind.TaggedTemplateExpression
+    | ts.SyntaxKind.ImportType
+    | ts.SyntaxKind.ExpressionWithTypeArguments
+    | ts.SyntaxKind.TypeReference
+    | ts.SyntaxKind.JsxOpeningElement
+    | ts.SyntaxKind.JsxSelfClosingElement
+    | ts.SyntaxKind.NewExpression
+    | ts.SyntaxKind.CallExpression;
+  [AST_NODE_TYPES.TSTypePredicate]: ts.SyntaxKind.TypePredicate;
+  [AST_NODE_TYPES.TSTypeQuery]: ts.SyntaxKind.TypeQuery;
+  [AST_NODE_TYPES.TSTypeReference]: ts.SyntaxKind.TypeReference;
+  [AST_NODE_TYPES.TSUndefinedKeyword]: ts.SyntaxKind.UndefinedKeyword;
+  [AST_NODE_TYPES.TSUnionType]: ts.SyntaxKind.UnionType;
+  [AST_NODE_TYPES.TSUnknownKeyword]: ts.SyntaxKind.UnknownKeyword;
+  [AST_NODE_TYPES.TSVoidKeyword]: ts.SyntaxKind.VoidKeyword;
+  [AST_NODE_TYPES.UpdateExpression]:
+    | ts.SyntaxKind.PrefixUnaryExpression
+    | ts.SyntaxKind.PostfixUnaryExpression;
+  [AST_NODE_TYPES.UnaryExpression]:
+    | ts.SyntaxKind.PrefixUnaryExpression
+    | ts.SyntaxKind.PostfixUnaryExpression
+    | ts.SyntaxKind.DeleteExpression
+    | ts.SyntaxKind.VoidExpression
+    | ts.SyntaxKind.TypeOfExpression;
+  [AST_NODE_TYPES.VariableDeclaration]:
+    | ts.SyntaxKind.VariableDeclarationList
+    | ts.SyntaxKind.VariableStatement;
+  [AST_NODE_TYPES.VariableDeclarator]: ts.SyntaxKind.VariableDeclaration;
+  [AST_NODE_TYPES.WhileStatement]: ts.SyntaxKind.WhileStatement;
+  [AST_NODE_TYPES.WithStatement]: ts.SyntaxKind.WithStatement;
+  [AST_NODE_TYPES.YieldExpression]: ts.SyntaxKind.YieldExpression;
+}
+
+export type TSESTreeToTSNode<T extends Node = Node> = Extract<
```

Comment:
Considering we have valid use cases for exposing this, we should add a comment to it.

```suggestion
/**
 * Maps TSESTree AST Node type to the expected TypeScript AST Node type(s).
 * This mapping is based on the internal logic of the parser.
 */
export type TSESTreeToTSNode<T extends Node = Node> = Extract<
```